lib/trino-parquet/src/main/java/io/trino/parquet/BloomFilterStore.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.log.Logger;
+import org.apache.parquet.column.values.bloomfilter.BlockSplitBloomFilter;
+import org.apache.parquet.column.values.bloomfilter.BloomFilter;
+import org.apache.parquet.format.BloomFilterHeader;
+import org.apache.parquet.format.Util;
+import org.apache.parquet.hadoop.metadata.ColumnChunkMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.io.SeekableInputStream;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class BloomFilterStore
+{
+    private static final Logger LOG = Logger.get(BloomFilterStore.class);
+
+    private final Map<ColumnPath, BloomFilter> cache = new HashMap<>();
+    private final ParquetDataSource dataSource;
+
+    @VisibleForTesting
+    BloomFilterStore()
+    {
+        this.dataSource = null;
+    }
+
+    public BloomFilterStore(ParquetDataSource dataSource)
+    {
+        this.dataSource = dataSource;
+    }
+
+    @VisibleForTesting
+    void insertBloomFilter(ColumnPath columnPath, BloomFilter bloomFilter)
+    {
+        cache.put(columnPath, bloomFilter);
+    }
+
+    public Optional<BloomFilter> readBloomFilter(ColumnChunkMetaData columnMetaData)
+    {
+        if (!bloomFilterEnabled(columnMetaData)) {
+            return Optional.empty();
+        }
+
+        if (cache.containsKey(columnMetaData.getPath())) {
+            return Optional.ofNullable(cache.get(columnMetaData.getPath()));
+        }
+
+        try {
+            Optional<BloomFilter> bloomFilterOptional = readBloomFilterInternal(columnMetaData);
+            if (bloomFilterOptional.isEmpty()) {
+                cache.put(columnMetaData.getPath(), null);
+            }
+            else {
+                cache.put(columnMetaData.getPath(), bloomFilterOptional.get());
+            }
+            return bloomFilterOptional;
+        }
+        catch (IOException exception) {
+            LOG.error(exception, "Failed to read Bloom filter data");
+        }
+
+        return Optional.empty();
+    }
+
+    public static boolean bloomFilterEnabled(ColumnChunkMetaData columnMetaData)
+    {
+        return columnMetaData.getBloomFilterOffset() > 0;
+    }
+
+    private Optional<BloomFilter> readBloomFilterInternal(ColumnChunkMetaData columnMetaData)
+            throws IOException
+    {
+        if (!bloomFilterEnabled(columnMetaData)) {
+            return Optional.empty();
+        }
+
+        long offset = columnMetaData.getBloomFilterOffset();
+        SeekableInputStream seekableInputStream = dataSource.seekableInputStream();
+        seekableInputStream.seek(offset);
```

Comment:
Using streaming read here seems to go against the existing norm of using positioned reads (https://trino.io/blog/2019/05/06/faster-s3-reads.html)
We could try doing this with positioned reads as described in https://github.com/trinodb/trino/pull/14202#issuecomment-1259330709
cc: @electrum

Reply:
Major changes:
load bloomfilter statistics based on https://gist.github.com/leetcode-1533/2fb1cf64d386c5bef4c26f4f37c9c714. (Via a two rounds of planRead)

I didn't do "We can derive a length for the read based on whether the bloom filter offset is less than start of 1st row group or greater than end of last row group.
In the former case we can read upto starts of the selected row groups, in the latter we can read upto start of page indexes or footer (which comes first).
So this is doable without resorting to streaming read APIs."

This is due to the first case (bloom filter stored before start of every row group) does not has an implementation and can't test the correctness of offset calculation.
The implementation didn't assume where the offset is, it assumed the relative small size of bloom filter header and read in those headers with a fixed amount of memory.

Please let me know if it is an issue.

[@el](cc: @electrum)